Dark and variance map generation from wx

zacsimile opened the pull request:
Response to Issue #140.

zacsimile committed:
Dark and variance map generation from wx

David-Baddeley commented on PYME/recipes/processing.py:
```diff
@@ -1959,3 +1959,103 @@ def execute(self, namespace):
+        from PYME.IO.MetaDataHandler import NestedClassMDHandler
+        from PYME.Analysis import gen_sCMOS_maps
+
+        logging.basicConfig() # without it got 'No handlers could be found for logger...'
```
not safe here - will override any previous logging config. This could be a bug if `dh5view` doesn't setup logging.

David-Baddeley replied:
Actually I see it's copied from the `main()` function in gen_scmos_maps.py, it makes sense in the main function of a standalone program, but shouldn't be copied over to a library.

David-Baddeley commented on PYME/recipes/processing.py:
```diff
@@ -1959,3 +1959,103 @@ def execute(self, namespace):
+    end = Int(0)
+    uniform = Bool(False)  # simulate a uniform map
+
+    def execute(self, namespace):
```
In general, recipe modules should be relatively light weight (i.e. the bulk of the logic should ideally reside elsewhere). As this is already in `gen_scmos_maps`, copying it here seems a bit odd.

David-Baddeley commented on PYME/recipes/processing.py:
```diff
@@ -1959,3 +1959,103 @@ def execute(self, namespace):
+
+        logging.basicConfig() # without it got 'No handlers could be found for logger...'
+
+        default_sensor_size = (2048,2048) # we currently assume this is correct but could be chosen based
```
If we are revisiting this, we should update this to use metadata entries (`"Camera.SensorWidth"`, `"Camera.SensorHeight"`)

David-Baddeley replied:
Oops, see that this is done below, ordering is a bit iffy.

David-Baddeley commented on PYME/recipes/processing.py:
```diff
@@ -1959,3 +1959,103 @@ def execute(self, namespace):
+        mfull, vefull, base_mdh = gen_sCMOS_maps.insert_into_full_map(m, ve, image.mdh, sensor_size=sensor_size)
+
+        common_md = NestedClassMDHandler()
+        common_md.setEntry('Analysis.name', 'mean-variance')
```
Need to fix these metadata entries to be vaguely sensible (they should not be in "Analysis" for a start)

David-Baddeley commented:
Unfortunately this isn't as easy as it might seem as `gen_sCMOS_maps.py` really needs some attention (it was a contribution from Christian which I merged in a hurry a long while back when trying to get the Exeter and main forks consistent. I'd always meant to revisit it but never got around to it.)

Potential issues:
- units need to be double checked (I think fitting might expect ADUs for variance, rather than electrons). This is now a largish problem as I think Andrew and I have been through and propagated various other units and documentation changes based on what was in gen_sCMOS_maps, rather than what the fitting actually wanted).
- some more general code style stuff.

David-Baddeley commented:
Just double checked units. The ORCA is currently writing it's metadata wrong, everything else, including the camera map generation seems OK. There are still some documentation issues (e.g. metadata docs).

David-Baddeley commented:
see #155

David-Baddeley commented:
@Zach - I'm doing a bit of tidying to gen_sCMOS_maps.py so hold fire on any modifications to this for now.

David-Baddeley committed:
Update __init__.py

camera maps definitely don't belong in the 'lite' set of modules, could potentially be added to base, but also OK as load on demand (or added to 'LM' personality).

David-Baddeley committed:
Use refactored gen_sCMOS_maps from GUI

Could optionally go back to using a recipe module once the recipe module has been re-written, but this works for now, and recipe would likely only add complexity.

David-Baddeley commented on PYME/recipes/processing.py:
```diff
@@ -1959,3 +1959,103 @@ def execute(self, namespace):
         namespace[self.output_name] = res
+
+@register_module('DarkAndVarianceMap')
+class DarkAndVarianceMap(ModuleBase):
```
Recipe module should call the refactored code in gen_sCMOS_maps (otherwise we have duplication). 

Do we need a recipe module? To what extent is this something we need to automate? For the standard full frame maps you pretty much need to take them once during the lifetime of a camera. This might change with maps for very fast frame rates on very small ROIs so I'm open to the idea of a recipe module.

David-Baddeley commented on PYME/recipes/processing.py:
```diff
@@ -1959,3 +1959,103 @@ def execute(self, namespace):
+    variance_threshold = Float(90000)  # again this is currently picked fairly arbitrarily
+    start = Int(0)
+    end = Int(0)
+    uniform = Bool(False)  # simulate a uniform map
```
If we do go with a recipe module, it should definitely not support uniform maps (the use case for these is very limited - namely testing equivalence of the map and non-map code, and I think it's reasonable that the user should have to jump through hoops to generate them).
